src/common/imageio_rawspeed.cc

```diff
@@ -333,16 +334,43 @@ dt_imageio_retval_t dt_imageio_open_rawspeed(dt_image_t *img, const char *filena
      * (from Klaus: r->pitch may differ from DT pitch (line to line spacing))
      * else fallback to generic dt_imageio_flip_buffers()
      */
-    const size_t bufSize_mipmap = (size_t)img->width * img->height * r->getBpp();
-    const size_t bufSize_rawspeed = (size_t)r->pitch * dimUncropped.y;
-    if(bufSize_mipmap == bufSize_rawspeed)
+
+    /*
+      For dual pixels we can't just copy or use the flip as data are interwoven.
+      Here we use the medium of both sub-pixels, not worth to use _OPENMP.
+      As there are no dual-float-pixel images we have them disabled.
+    */
+    if (cpp == 1)
     {
-      memcpy(buf, r->getDataUncropped(0, 0), bufSize_mipmap);
+      const size_t bufSize_mipmap = (size_t)img->width * img->height * r->getBpp();
+      const size_t bufSize_rawspeed = (size_t)r->pitch * dimUncropped.y;
+      if(bufSize_mipmap == bufSize_rawspeed)
+      {
+        memcpy(buf, r->getDataUncropped(0, 0), bufSize_mipmap);
+      }
+      else
+      {
+        dt_imageio_flip_buffers((char *)buf, (char *)r->getDataUncropped(0, 0), r->getBpp(), dimUncropped.x,
+                                dimUncropped.y, dimUncropped.x, dimUncropped.y, r->pitch, ORIENTATION_NONE);
+      }
     }
     else
     {
-      dt_imageio_flip_buffers((char *)buf, (char *)r->getDataUncropped(0, 0), r->getBpp(), dimUncropped.x,
-                              dimUncropped.y, dimUncropped.x, dimUncropped.y, r->pitch, ORIENTATION_NONE);
+      if(img->buf_dsc.datatype == TYPE_UINT16)
+      {
+        for(int j = 0; j < img->height; j++)
+        {
+          const uint16_t *in = (uint16_t *) r->getDataUncropped(0, j);
+          uint16_t *out = ((uint16_t *)buf) + (size_t) j * img->width;
+          for(int i = 0; i < img->width; i++)
+            out[i] = (in[i*2] + in[i*2+1]) / 2;
```

Comment:
1. This doesn't really look like correct dual pixel handling - is averaging really the correct handling?
2. how do we know this handling is correct for every cpp=2 file out there?

i suspect it should be an iop

Reply:
I think you are right here in a general sense. For these canon dual pixels it's correct but might be completely wrong in other cases.

Reply:
> i suspect it should be an iop

I don't understand ...

Reply:
I don't know what darktable does internally nowadays, but it really doesn't look correct to do
such an operation *here*, not in the pixel pipeline, if requested by some module (iop).

Reply:
Aah. This pr is **not** intended as something clever to do with dual-pixel images, for that we would of course need to pass all data of the images in an ordered manner and make access to that data valid in the pixel pipeline. That would require much more work and changes and i doubt it's really worth it. (Personally i absolutely do not care about those dual pixel files). And so far none of the devs ever seemed to need/implement this feature... missing something maybe...

This pr **only** presents the raw data of one of the images so the rest of the pipeline can access that data as if there was only one. Thus i thought it would be ok this way as we do the copy-data only once when reading into the cache.

At least people can use these dng files as if they were just-raw only missing the 'candy-features' of dual-pixel.

Reply:
I wonder if user has to be informed somehow that this is not a real dual pixel support.

Reply:
In release note or as a hint message when opening?

Reply:
dunno... At least one of those;)

Reply:
That didn't really answer my second question.
Even if it results in not-totally-terrible combined images for these dual-pixel raws,

> 2. how do we know this handling is correct for every cpp=2 file out there?

Reply:
1. They are not _not-totally-terrible combined images_, what we see here is exactly one of two images, what we don't is some clever algorithm combining both.

2. An yes, i am not totally sure about **every** file, unfortunately there is no valid documentation i found so far. Do you know of any?